False positive for undefined variable with assignment expression

### Steps to reproduce
```
~/tmp λ cat test4.py
def i(am: str = (the := "walrus")) -> str:
    return "goo goo g'joob"
print(the)
```
### Current behavior
```
~/tmp λ pylint -E test4.py
************* Module test4
test4.py:5:6: E0602: Undefined variable 'the' (undefined-variable)
```
### Expected behavior
pylint recognises that `the` is defined. Alternatively, pylint yells at me for doing this in the first place :-)

### pylint --version output
```
~/tmp λ pylint --version
pylint 2.5.3
astroid 2.4.2
Python 3.8.3 (default, May 27 2020, 20:54:22) 
[Clang 11.0.3 (clang-1103.0.32.59)]
```

Pretty minor, especially given that assignment expressions aren't a priority for pylint (https://github.com/PyCQA/pylint/issues/3275#issuecomment-559122266).

--- a/ChangeLog
+++ b/ChangeLog
@@ -47,6 +47,12 @@ Release date: TBA
 
   Fixes part of #3688
 
+* ``undefined-variable`` now correctly triggers for assignment expressions in if ... else statements
+  This includes a basic form of control flow inference for if ... else statements using
+  constant boolean values
+
+  Closes #3688
+
 * Fix bug with importing namespace packages with relative imports
 
   Closes #2967 and #5131

--- a/doc/whatsnew/2.12.rst
+++ b/doc/whatsnew/2.12.rst
@@ -96,6 +96,12 @@ Other Changes
 
   Fixes part of #3688
 
+* ``undefined-variable`` now correctly triggers for assignment expressions in if ... else statements
+  This includes a basic form of control flow inference for if ... else statements using
+  constant boolean values
+
+  Closes #3688
+
 * Fix double emitting of ``not-callable`` on inferrable ``properties``
 
   Closes #4426

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1211,6 +1211,12 @@ class VariablesChecker(BaseChecker):
                         )
                     if is_first_level_ref:
                         break
+                elif isinstance(defnode, nodes.NamedExpr):
+                    if isinstance(defnode.parent, nodes.IfExp):
+                        if self._is_never_evaluated(defnode, defnode.parent):
+                            self.add_message(
+                                "undefined-variable", node=node, args=node.name
+                            )
 
             current_consumer.mark_as_consumed(node.name, found_nodes)
             # check it's not a loop variable used outside the loop
@@ -1634,6 +1640,21 @@ class VariablesChecker(BaseChecker):
             ):
                 return 2
         return 0
+
+    @staticmethod
+    def _is_never_evaluated(
+        defnode: nodes.NamedExpr, defnode_parent: nodes.IfExp
+    ) -> bool:
+        """Check if a NamedExpr is inside a side of if ... else that never
+        gets evaluated
+        """
+        inferred_test = utils.safe_infer(defnode_parent.test)
+        if isinstance(inferred_test, nodes.Const):
+            if inferred_test.value is True and defnode == defnode_parent.orelse:
+                return True
+            if inferred_test.value is False and defnode == defnode_parent.body:
+                return True
+        return False
 
     def _ignore_class_scope(self, node):
         """

--- a/tests/functional/u/undefined/undefined_variable_py38.py
+++ b/tests/functional/u/undefined/undefined_variable_py38.py
@@ -1,5 +1,5 @@
 """Tests for undefined variable with assignment expressions"""
-# pylint: disable=using-constant-test
+# pylint: disable=using-constant-test, expression-not-assigned
 
 # Tests for annotation of variables and potentially undefinition
 
@@ -48,3 +48,52 @@ def no_parameters_in_function_default() -> None:
 
 
 print(again_no_default) # [undefined-variable]
+
+# Tests for assignment expressions in if ... else comprehensions
+
+
+[i for i in range(10) if (if_assign_1 := i)]
+
+print(if_assign_1)
+
+IF_TWO = [i for i in range(10) if (if_assign_2 := i)]
+
+print(if_assign_2)
+
+IF_THREE = next(i for i in range(10) if (if_assign_3 := i))
+
+print(if_assign_3)
+
+IF_FOUR = {i: i for i in range(10) if (if_assign_4 := i)}
+
+print(if_assign_4)
+
+IF_FIVE = {i: i if (if_assign_5 := i) else 0 for i in range(10)}
+print(if_assign_5)
+
+{i: i if True else (else_assign_1 := i) for i in range(10)}
+
+print(else_assign_1) # [undefined-variable]
+
+
+# Tests for assignment expressions in the assignment of comprehensions
+
+[(assign_assign_1 := i) for i in range(10)]
+
+print(assign_assign_1)
+
+COMPREHENSION_TWO =[(assign_assign_2 := i) for i in range(10)]
+
+print(assign_assign_2)
+
+COMPREHENSION_THREE = next((assign_assign_3 := i) for i in range(10))
+
+print(assign_assign_3)
+
+COMPREHENSION_FOUR = {i: (assign_assign_4 := i) for i in range(10)}
+
+print(assign_assign_4)
+
+COMPREHENSION_FIVE = {i: (else_assign_2 := i) if False else 0 for i in range(10)}
+
+print(else_assign_2)  # [undefined-variable]

--- a/tests/functional/u/undefined/undefined_variable_py38.txt
+++ b/tests/functional/u/undefined/undefined_variable_py38.txt
@@ -1,3 +1,5 @@
 undefined-variable:17:15:typing_and_self_referncing_assignment_expression:Undefined variable 'var':HIGH
 undefined-variable:42:6::Undefined variable 'no_default':HIGH
 undefined-variable:50:6::Undefined variable 'again_no_default':HIGH
+undefined-variable:76:6::Undefined variable 'else_assign_1':HIGH
+undefined-variable:99:6::Undefined variable 'else_assign_2':HIGH

